parquet-column/src/main/java/org/apache/parquet/column/values/bloomfilter/BlockSplitBloomFilter.java

```diff
@@ -0,0 +1,372 @@
+    }
+  }
+
+  @Override
+  public boolean findHash(long hash) {
+    long numBlocks = bitset.length / BYTES_PER_BLOCK;
+    long lowHash = hash >>> 32;
+    int blockIndex = (int)((lowHash * numBlocks) >> 32);
+    int key = (int)hash;
+
+    // Calculate mask for the tiny Bloom filter.
+    int[] mask = setMask(key);
+    for (int i = 0; i < BITS_SET_PER_BLOCK; i++) {
+      if (0 == (intBuffer.get(blockIndex * (BYTES_PER_BLOCK / 4) + i) & mask[i])) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Calculate optimal size according to the number of distinct values and false positive probability.
+   *
+   * @param n: The number of distinct values.
+   * @param p: The false positive probability.
+   *
+   * @return optimal number of bits of given n and p.
+   */
+  public static int optimalNumOfBits(long n, double p) {
+    Preconditions.checkArgument((p > 0.0 && p < 1.0),
+      "FPP should be less than 1.0 and great than 0.0");
+    final double m = -8 * n / Math.log(1 - Math.pow(p, 1.0 / 8));
+    final double MAX = UPPER_BOUND_BYTES << 3;
+    int numBits = (int)m;
+
+    // Handle overflow.
+    if (m > MAX || m < 0) {
+      numBits = (int)MAX;
+    }
+
+    // Round to BITS_PER_BLOCK
+    numBits = (numBits + BITS_PER_BLOCK -1) & ~BITS_PER_BLOCK;
+
+    if (numBits < (LOWER_BOUND_BYTES << 3)) {
+      numBits = LOWER_BOUND_BYTES << 3;
+    }
+
+    return numBits;
+  }
+
+  @Override
+  public int getBitsetSize() {
+    return this.bitset.length;
+  }
+
+  @Override
+  public long hash(Object value) {
+    ByteBuffer plain;
+
+    if (value instanceof Binary) {
+      return hashFunction.hashBytes(((Binary) value).getBytes());
+    }
+
+    if (value instanceof Integer) {
+      plain = ByteBuffer.allocate(Integer.SIZE/Byte.SIZE);
+      plain.order(ByteOrder.LITTLE_ENDIAN).putInt(((Integer)value));
+    } else if (value instanceof Long) {
+      plain = ByteBuffer.allocate(Long.SIZE/Byte.SIZE);
+      plain.order(ByteOrder.LITTLE_ENDIAN).putLong(((Long)value));
+    } else if (value instanceof Float) {
+      plain = ByteBuffer.allocate(Float.SIZE/Byte.SIZE);
+      plain.order(ByteOrder.LITTLE_ENDIAN).putFloat(((Float)value));
+    } else if (value instanceof Double) {
+      plain = ByteBuffer.allocate(Double.SIZE/ Byte.SIZE);
+      plain.order(ByteOrder.LITTLE_ENDIAN).putDouble(((Double)value));
+    } else {
+      throw new RuntimeException("Parquet Bloom filter: Not supported type");
+    }
+
+    return hashFunction.hashByteBuffer(plain);
+  }
+
+  @Override
+  public HashStrategy getHashStrategy() {
+    return HashStrategy.XXH64;
+  }
+
+  @Override
+  public Algorithm getAlgorithm() {
+    return Algorithm.BLOCK;
+  }
+
+  @Override
+  public Compression getCompression() {
+    return Compression.UNCOMPRESSED;
+  }
+
+  @Override
+  public long hash(int value) {
+    ByteBuffer plain = ByteBuffer.allocate(Integer.SIZE/Byte.SIZE);
```

Comment:
I agree to use a class level ByteBuffer to speed up.  The reason for using hashByteBuffer is to avoid endianness issue, though LongHashFunction seems to use little-endian by default. Let me check the xxhash implementation again. I will put comments in the code to address this.